crepecake/src/integration-test/java/com/google/cloud/tools/crepecake/registry/ManifestPullerIntegrationTest.java

```diff
@@ -0,0 +1,122 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.registry;
+
+import com.google.cloud.tools.crepecake.image.json.ManifestTemplateHolder;
+import com.google.cloud.tools.crepecake.image.json.UnknownManifestFormatException;
+import com.google.cloud.tools.crepecake.image.json.V21ManifestTemplate;
+import com.google.cloud.tools.crepecake.image.json.V22ManifestTemplate;
+import java.io.IOException;
+import java.util.logging.Logger;
+import org.hamcrest.CoreMatchers;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class ManifestPullerIntegrationTest {
+
+  private static final Logger logger =
+      Logger.getLogger(ManifestPullerIntegrationTest.class.getName());
+
+  /**
+   * True if the local registry is running. These tests are skipped if the local registry failed to
+   * start.
+   */
+  private static boolean localRegistryRunning;
+
+  @BeforeClass
+  public static void startLocalRegistry() {
+    try {
+      String runRegistryCommand =
+          "docker run -d -p 5000:5000 --restart=always --name registry registry:2";
+      Runtime.getRuntime().exec(runRegistryCommand).waitFor();
+
+      String pullImageCommand = "docker pull busybox";
+      Runtime.getRuntime().exec(pullImageCommand).waitFor();
+
+      String tagImageCommand = "docker tag busybox localhost:5000/busybox";
+      Runtime.getRuntime().exec(tagImageCommand).waitFor();
+
+      String pushImageCommand = "docker push localhost:5000/busybox";
+      Runtime.getRuntime().exec(pushImageCommand).waitFor();
+
+      localRegistryRunning = true;
+    } catch (IOException | InterruptedException ex) {
+      // The local registry failed to start.
+      logger.info(
+          "Skipping ManifestPullerIntegrationTest because failed to set up local registry: "
+              + ex.getMessage());
+      localRegistryRunning = false;
+    }
+  }
+
+  @AfterClass
+  public static void stopLocalRegistry() throws IOException, InterruptedException {
+    if (!localRegistryRunning) {
+      return;
+    }
+
+    String stopRegistryCommand = "docker stop registry";
+    Runtime.getRuntime().exec(stopRegistryCommand).waitFor();
+
+    String removeRegistryContainerCommand = "docker rm -v registry";
+    Runtime.getRuntime().exec(removeRegistryContainerCommand).waitFor();
+  }
+
+  @Test
+  public void testPull()
+      throws IOException, RegistryErrorException, RegistryUnauthorizedException,
+          RegistryTooManyRequestsException, UnknownManifestFormatException {
+    if (!localRegistryRunning) {
+      return;
+    }
+
+    ManifestPuller manifestPuller = new ManifestPuller(null, "localhost:5000", "busybox");
+    ManifestTemplateHolder manifestTemplateHolder = manifestPuller.pull("latest");
+
+    if (manifestTemplateHolder.isV21()) {
```

Comment:
Conditional logic in unit tests is suspect. One or the other of these should be true. Just test that.